feat(format): introduce ADBC API revision 1.1.0

lidavidm opened the pull request:
Fixes #317.

lidavidm commented on adbc.h:
```diff
@@ -667,6 +675,34 @@ struct ADBC_EXPORT AdbcDriver {
+  ///
+  /// @{
+
+  /// Pad the struct to have 64 pointers.  Space reserved for future
+  /// growth.
+  void* reserved[35];
```
I'm going to add everything in this revision. I think what I'll actually do now is update this PR with all the spec changes for C/C++ but without implementations, then follow up with one PR per spec change for implementation (ditto for Java, Go). And we can tweak the spec changes if we find something during implementation that doesn't mesh.

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -640,8 +640,8 @@ AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
 
-  if (version != ADBC_VERSION_1_0_0) {
-    SetError(error, "Only ADBC 1.0.0 is supported");
+  if (version != ADBC_VERSION_1_0_0 && version != ADBC_VERSION_1_1_0) {
+    SetError(error, "Only ADBC 1.0.0 and 1.1.0 are supported");
     return ADBC_STATUS_NOT_IMPLEMENTED;
   }
```
to avoid this becoming a huge thing in the future, should we make this a switch with a default? that way future updates would just be to add a new case statement to it rather than this if condition just growing ever larger

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -782,11 +782,14 @@ AdbcStatusCode AdbcLoadDriverFromInitFunc(AdbcDriverInitFunc init_func, int vers
   }
 
   auto result = init_func(version, raw_driver, error);
+  if (version == ADBC_VERSION_1_1_0 && result == ADBC_STATUS_NOT_IMPLEMENTED) {
+    result = init_func(ADBC_VERSION_1_0_0, raw_driver, error);
+  }
```
maybe create a static const list of versions and just go through them in order until we get something other than `ADBC_STATUS_NOT_IMPLEMENTED` calling `init_func`? That way future updates are just adding the version to the list?

lidavidm commented:
I've addressed feedback + added APIs for most of the new proposals. If this is OK I'll follow up with Go (with a set of interfaces, one per method or so) and Java (which is very easy).

I think we will have to punt on #320 and #621. I don't think #630 is worth doing. #685 is too speculative at this point.

zeroshade commented on adbc.h:
```diff
@@ -1138,6 +1293,43 @@ AdbcStatusCode AdbcStatementBindStream(struct AdbcStatement* statement,
+AdbcStatusCode AdbcStatementCancel(struct AdbcStatement* statement,
+                                   struct AdbcError* error);
+
+/// \brief Get a double property of the statement.
```
what would this be used for? is it just to avoid consumers having to call `strtof`/`strtod` themselves?

lidavidm replied:
There's a constant above motivating it. Also, 1) there's no Get* function at all right now and 2) returning strings is really annoying due to allocations. 

I could change this to return a string, the caveats would be either: 1) we have to add some release-callback thing (C Data Interface-ish) or 2) we return a const char*, _but_ calling this again or doing other things with the statement will invalidate the pointer (SQLite-ish) or 3) we expect a caller-allocated and/or fixed-size buffer (ODBC-ish)

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -816,6 +846,13 @@ AdbcStatusCode AdbcLoadDriverFromInitFunc(AdbcDriverInitFunc init_func, int vers
     FILL_DEFAULT(driver, StatementSetSqlQuery);
     FILL_DEFAULT(driver, StatementSetSubstraitPlan);
   }
+  if (version >= ADBC_VERSION_1_1_0) {
+    auto* driver = reinterpret_cast<struct AdbcDriver*>(raw_driver);
+    FILL_DEFAULT(driver, StatementExecuteSchema);
```
shouldn't this also have the same for the other new functions? `AdbcStatementCancel` and `AdbcStatementGetDouble`?

lidavidm replied:
I'm going to defer implementation, just wanted to demonstrate the basic idea worked. I just want to get the API sort of in the right shape.

lidavidm replied:
Agh, let me retarget this PR against the branch, not main.

lidavidm replied:
There - the intent is to build this out on a branch then manually ff-merge it at the end. If the approach looks reasonable I'm also going to build out more serious compatibility tests.

zeroshade replied:
Gotcha, okay then in that case I'm in favor of this approach.

lidavidm commented:
Interestingly the CI failure exposes what does look like a compatibility issue: the new fields in AdbcDriver make the Python test crash (because it mixes a driver using the new definition and a wheel using the old definition)

lidavidm commented:
Ah, because the drivers do this

```
  auto* driver = reinterpret_cast<struct AdbcDriver*>(raw_driver);
  std::memset(driver, 0, sizeof(*driver));
```

and now that the struct has changed, they need to actually check the size of the struct they're about to manipulate.

lidavidm commented:
I'm quite tempted to just remove cpplint, it seems to mostly just give noise and it's basically unconfigurable/unmaintained

lidavidm commented:
Some other thoughts: maybe we shouldn't overload getInfo for things like current catalog/schema, and instead add a get/set option pair? That would be consistent with how Java does this

lidavidm commented:
Added Java and Go, though I think I'd like to go back and align things a bit.

lidavidm commented:
There's a macOS error in CI that I'll try to track down, but otherwise I think this is ready (for C, Java, Go)

lidavidm commented:
TODOs

- [ ] Reserve more space for AdbcDriver?
- [ ] Figure out macOS CI error

lidavidm commented:
I never really learned amd64 assembly but I think the CI failure I'm seeing is possibly a miscompilation or something odd? I see basically an unconditional call to UBSAN's abort here:

```
    0x103770820 <+64>:   movq   -0x10(%rbp), %rax
    0x103770824 <+68>:   movq   %rax, -0x20(%rbp)
    0x103770828 <+72>:   movq   -0x20(%rbp), %rax
    0x10377082c <+76>:   movq   %rax, -0x28(%rbp)
    0x103770830 <+80>:   xorl   %eax, %eax
    0x103770832 <+82>:   testb  $0x1, %al
    0x103770834 <+84>:   jne    0x10377084a               ; <+106> at postgresql.cc
    0x10377083a <+90>:   leaq   0x109f3f(%rip), %rdi      ; _dyld_private + 158976
    0x103770841 <+97>:   xorl   %eax, %eax
    0x103770843 <+99>:   movl   %eax, %esi
    0x103770845 <+101>:  callq  0x1037f75ae               ; symbol stub for: __ubsan_handle_type_mismatch_v1_abort
```

aka (if I'm reading it right):
- -0x10(%rbp) is `raw_driver`,
- 4 instructions of shuffling registers/stack,
- a comparison that always fails (since we zero out EAX, then bitwise-AND it with 1 before doing a conditional jump),
- an unconditional call to `__ubsan_handle_type_mismatch_v1_abort`

Ghidra seems to agree:

```c
  if (param_1 == 1000000) {
    if (param_2 == 0) {
      local_9 = 5;
    }
    else {
      ___ubsan_handle_type_mismatch_v1_abort(&DAT_00194780,0);
```

I'm not sure why this is happening, or if there's UB in the code and this is a consequence of it...

In the debugger, I can confirm `raw_driver` is not NULL despite UBSAN reporting a null-pointer-dereference.

lidavidm commented:
For now let's see if writing it differently helps...

lidavidm commented:
Alright, CI passes here: https://github.com/lidavidm/arrow-adbc/tree/spec-1.1.0-newstyle

I'll go ahead and start working on implementation in a new branch while leaving this for review.

kou reviewed (approved):
+1

lidavidm commented on adbc.h:
```diff
@@ -1138,6 +1549,101 @@ AdbcStatusCode AdbcStatementBindStream(struct AdbcStatement* statement,
+/// This can be called during AdbcStatementExecuteQuery (or similar),
+/// or while consuming an ArrowArrayStream returned from such.
+/// Calling this function should make the other functions return
+/// ADBC_STATUS_CANCELLED or ECANCELED (for ArrowArrayStream).
```
ArrowArrayStream's methods return an `int` that is specified to be `errno`-compatible, but not actually errno:

https://github.com/apache/arrow/blob/fbe5f641d327ee81db00ce5f056940a69f4d8603/cpp/src/arrow/c/abi.h#L75

lidavidm replied:
I clarified this a bit.

kou replied:
Ah, I see.

lidavidm commented on adbc.h:
```diff
@@ -694,6 +961,24 @@ ADBC_EXPORT
+///
+/// \since ADBC API revision 1.1.0
+/// \addtogroup adbc-1.1.0
+/// \return ADBC_STATUS_NOT_IMPLEMENTED if the option is not recognized
```
Done!

lidavidm commented:
I will have to come back and add support for [ArrowDeviceArray](https://lists.apache.org/thread/o2hsw7o1gm3qgw5z51rmz6zqxh0p7bvk) once that vote passes. And probably, some standard options to hint to drivers to try to use a particular device if possible. (I think this is basically only possible for Flight-UCX, or maybe for a hypothetical database that uses an accelerator and is embeddable, like the GPU analog to SQLite.)

lidavidm commented:
Ah, and right now we can't even support Flight-UCX since we aren't using the C++ Flight implementation anyways.

lidavidm commented:
I added language clarifying that returned ArrowArrayStreams must outlive the object that created them, since that seemed to be a complaint in #702

pitrou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -781,12 +810,17 @@ AdbcStatusCode AdbcLoadDriverFromInitFunc(AdbcDriverInitFunc init_func, int vers
-  auto result = init_func(version, raw_driver, error);
+  AdbcStatusCode result = ADBC_STATUS_NOT_IMPLEMENTED;
+  for (const int try_version : kSupportedVersions) {
+    if (try_version > version) continue;
```
I'm trying to understand the logic. Does it mean that, if the user passes a future `ADBC_VERSION_1_2_0`, we try calling `init_func` with `ADBC_VERSION_1_1_0` and then with `ADBC_VERSION_1_0_0`?

pitrou replied:
Perhaps you should explain the intent in a comment.

lidavidm replied:
Right, so as mentioned above this should check for a supported version first. I've clarified in the source: the intent is that if they pass 1_1_0, we'll try to pass that to the underlying driver, then try 1_0_0 with the driver. If they pass 1_0_0 we'll only try 1_0_0.

pitrou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -640,11 +656,19 @@ AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
+    case ADBC_VERSION_1_0_0:
+    case ADBC_VERSION_1_1_0:
+      break;
+    default:
+      SetError(error, "Only ADBC 1.0.0 and 1.1.0 are supported");
+      return ADBC_STATUS_NOT_IMPLEMENTED;
```
Why are being stricter than `AdbcLoadDriverFromInitFunc` here?

pitrou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -640,11 +656,19 @@ AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
+      return ADBC_STATUS_NOT_IMPLEMENTED;
   }
 
+  if (!raw_driver) {
+    SetError(error, "Must provide non-NULL raw_driver");
+    return ADBC_STATUS_INVALID_ARGUMENT;
+  }
```
Should this check be moved in `AdbcLoadDriverFromInitFunc`?

lidavidm replied:
I replicated the check there.

pitrou commented on adbc.h:
```diff
@@ -315,6 +354,16 @@ struct ADBC_EXPORT AdbcError {
 #define ADBC_INFO_DRIVER_ARROW_VERSION 102
+/// \brief The driver ADBC API version (type: int64).
+///
+/// The value should be one of the ADBC_VERSION constants.
```
Is it allowed to be a later version?

lidavidm replied:
"Yes", in that if/when there's another revision, then that would be a valid value, but earlier clients/drivers of course wouldn't know about that.

pitrou commented on adbc.h:
```diff
@@ -667,8 +786,80 @@ struct ADBC_EXPORT AdbcDriver {
+  /// ADBC_VERSION_1_1_0.
+  ///
+  /// For a 1.0.0 driver being loaded by a 1.1.0 driver manager: the
+  /// 1.1.0 manager will allocate the new, expanded AdbcDriver struct
```
Does the driver manager allocate anything? In `AdbcLoadDriver` it's the caller that passes the `AdbcDriver` pointer.

lidavidm replied:
If you use the `AdbcDatabaseInit` path (where the driver manager pretends to just be an ADBC driver, and you pass the driver name as one of the init options) then it will allocate for you, yes.

pitrou commented on adbc.h:
```diff
@@ -667,8 +786,80 @@ struct ADBC_EXPORT AdbcDriver {
+  /// and attempt to have the driver initialize it with
+  /// ADBC_VERSION_1_1_0.  This must return an error, after which the
+  /// driver will try again with ADBC_VERSION_1_0_0.  The driver must
+  /// not access the new fields.
```
The new fields will be null-initialized?

lidavidm replied:
In this scenario, the new fields do not exist because it is an old driver manager working with a new driver.

pitrou replied:
Ok, so perhaps clarify things a bit? "The driver must not access the new fields, which will carry undefined values" perhaps?

pitrou commented on adbc.h:
```diff
@@ -667,8 +786,80 @@ struct ADBC_EXPORT AdbcDriver {
+///
+/// \since ADBC API revision 1.1.0
+/// \addtogroup adbc-1.1.0
+#define ADBC_DRIVER_1_0_0_SIZE (offsetof(struct AdbcDriver, DatabaseGetOption))
```
Should we also have `ADBC_DRIVER_1_1_0_SIZE`? That way, a 1.1.0 driver could perhaps be source-compatible with later versions of `adbc.h`.

lidavidm replied:
Added.

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -540,6 +547,15 @@ AdbcStatusCode AdbcStatementExecuteQuery(struct AdbcStatement* statement,
+  if (!statement->private_driver) {
+    return ADBC_STATUS_INVALID_STATE;
+  }
+  return statement->private_driver->StatementExecuteSchema(statement, schema, error);
```
It's assumed that if you're using these functions, you're using the driver manager, and that the driver manager has initialized all the functions. (Obviously, not true right now, since I decided to split implementation out.)

lidavidm commented on adbc.h:
```diff
@@ -667,8 +786,80 @@ struct ADBC_EXPORT AdbcDriver {
+  AdbcStatusCode (*StatementSetOptionDouble)(struct AdbcStatement*, const char*, double,
+                                             struct AdbcError*);
+
+  /// Pad the struct to have 96 pointers.  Space reserved for future growth.
```
It's not strictly necessary but would let us avoid technically breaking ABI in the future without having to be as careful about what's going on.

pitrou replied:
I don't know, to me it seems a bit weird to have two different ABI-stability mechanisms. It adds cognitive overload.

lidavidm commented:
I'm going to merge this soon (into the branch) and then update/continue the follow-up PRs.

lidavidm commented:
https://github.com/apache/arrow-adbc/pull/731
